src/ray/gcs/gcs_server/gcs_resource_scheduler.cc

```diff
@@ -0,0 +1,301 @@
+    result.clear();
+  }
+  return result;
+}
+
+std::vector<NodeID> GcsResourceScheduler::StrictPackSchedule(
+    const std::vector<ResourceSet> &required_resources_list,
+    const absl::flat_hash_set<NodeID> &candidate_nodes) {
+  std::vector<NodeID> result;
+
+  // Aggregate required resources.
+  ResourceSet required_resources;
+  for (const auto &iter : required_resources_list) {
+    required_resources.AddResources(iter);
+  }
+
+  // Score and sort nodes.
+  const auto &node_scores = ScoreNodes(required_resources, candidate_nodes);
+
+  // Select the node with the highest score.
+  // `StrictPackSchedule` does not need to consider the scheduling context, because it
+  // only schedules to a node and triggers rescheduling when node dead.
+  if (!node_scores.empty() && node_scores.front().second >= 0) {
+    for (int index = 0; index < (int)required_resources_list.size(); ++index) {
+      result.push_back(node_scores.front().first);
+    }
+  }
+  return result;
+}
+
+std::vector<NodeID> GcsResourceScheduler::PackSchedule(
+    const std::vector<ResourceSet> &required_resources_list,
+    const absl::flat_hash_set<NodeID> &candidate_nodes) {
+  std::vector<NodeID> result;
+  result.resize(required_resources_list.size());
+  absl::flat_hash_set<NodeID> candidate_nodes_copy(candidate_nodes);
+  std::list<std::pair<int, ResourceSet>> required_resources_list_copy;
+  int index = 0;
+  for (const auto &iter : required_resources_list) {
+    required_resources_list_copy.emplace_back(index++, iter);
+  }
+
+  while (!required_resources_list_copy.empty()) {
+    // Score and sort nodes.
+    const auto &required_resources_index = required_resources_list_copy.front().first;
+    const auto &required_resources = required_resources_list_copy.front().second;
+    const auto &node_scores = ScoreNodes(required_resources, candidate_nodes_copy);
+    if (node_scores.empty() || node_scores.front().second < 0) {
+      // There is no node to meet the scheduling requirements.
+      break;
+    }
+
+    const auto &highest_score_node_id = node_scores.front().first;
+    RAY_CHECK(gcs_resource_manager_.AcquireResources(highest_score_node_id,
+                                                     required_resources));
+    result[required_resources_index] = highest_score_node_id;
+    required_resources_list_copy.pop_front();
+
+    // We try to schedule more resources on one node.
+    for (auto iter = required_resources_list_copy.begin();
+         iter != required_resources_list_copy.end();) {
+      if (gcs_resource_manager_.AcquireResources(highest_score_node_id, iter->second)) {
+        result[iter->first] = highest_score_node_id;
+        required_resources_list_copy.erase(iter++);
+      } else {
+        ++iter;
+      }
+    }
+    candidate_nodes_copy.erase(highest_score_node_id);
+  }
+
+  // Releasing the resources temporarily deducted from `gcs_resource_manager_`.
+  ReleaseTemporarilyDeductedResources(required_resources_list, result);
```

Comment:
This code path should be unit tested to make sure it doesn't leak resources.

Reply:
I have add test case, thx.